src/main/java/com/microsoft/spring/data/gremlin/query/GremlinTemplate.java

```diff
@@ -349,44 +328,49 @@ public long edgeCount() {
         return results.size();
     }
 
-    private <T> T recoverDomain(@NonNull GremlinSource source, @NonNull Result result,
-                                @NonNull Class<T> domainClass, boolean isEntityEdge) {
+    private <T> T recoverDomain(@NonNull GremlinSource source, @NonNull Result result) {
         final T domain;
+        @SuppressWarnings("unchecked") final Class<T> domainClass = (Class<T>) source.getDomainClass();
 
         source.doGremlinResultRead(result);
         domain = this.mappingConverter.read(domainClass, source);
 
-        if (isEntityEdge) {
+        if (source instanceof GremlinSourceEdge) {
             this.completeEdge(domain, (GremlinSourceEdge) source);
         }
 
         return domain;
     }
 
-    private <T> List<T> recoverDomainList(@NonNull GremlinSource source, @NonNull List<Result> results,
-                                          @NonNull Class<T> domainClass, boolean isEntityEdge) {
-        final List<T> domainList = new ArrayList<>();
+    private <T> List<T> recoverDomainList(@NonNull GremlinSource source, @NonNull List<Result> results) {
```

Comment:
Same comments as for the `recoverDomain` method. `T` plays no purpose here, unless we have `GremlinSource<T>`